src/wrappers/nats.js

```diff
@@ -0,0 +1,205 @@
+            triggerMetadata.sid = callback_sid;
+        }
+        // Finalize nats event.
+        eventInterface.finalizeEvent(natsEvent, natsStartTime, null, triggerMetadata);
+        const { slsEvent: nodeEvent, startTime: nodeStartTime } = eventInterface.initializeEvent(
+            'node_function', 'messageHandler', 'messageReceived', 'runner'
+        );
+        try {
+            runnerResult = callback(callback_msg, callback_reply, callback_subject, callback_sid);
+        } catch (err) {
+            originalHandlerSyncErr = err;
+        }
+        const originalHandlerName = callback.name;
+        if (originalHandlerName) {
+            nodeEvent.getResource().setName(originalHandlerName);
+        }
+        // Handle and finalize async user function.
+        if (utils.isPromise(runnerResult)) {
+            let originalHandlerAsyncError;
+            runnerResult = runnerResult.catch((err) => {
+                originalHandlerAsyncError = err;
+                throw err;
+            }).finally(() => {
+                eventInterface.finalizeEvent(nodeEvent, nodeStartTime, originalHandlerAsyncError);
+                tracer.sendTrace(() => {});
+            });
+        } else {
+            // Finalize sync user function.
+            eventInterface.finalizeEvent(nodeEvent, nodeStartTime, originalHandlerSyncErr);
+            tracer.sendTrace(() => {});
+        }
+        tracer.addRunner(nodeEvent, runnerResult);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    // Throwing error in case of sync user function.
+    if (originalHandlerSyncErr) {
+        throw originalHandlerSyncErr;
+    }
+    return runnerResult;
+}
+
+/**
+ * Wraps nats subscribe function with tracing.
+ * @param {Function} wrappedFunction nats subscribe function.
+ * @param {string} serverHostname nats server host name.
+ * @return {Function} subscribe function rsoponse.
+ */
+function natsSubscribeWrapper(wrappedFunction, serverHostname) {
+    return function internalNatsSubscribeWrapper(subject, opts, callback) {
+        let clientRequest;
+        const { opts_internal, callback_internal } = getSubscribeParams(opts, callback);
+        let patchedCallback = callback_internal;
+        try {
+            const isRequestCall = isNatsRequestCall(subject);
+            patchedCallback = (callback_msg, callback_reply, callback_subject, callback_sid) => {
+                traceContext.RunInContext(
+                    tracer.createTracer,
+                    () => natsSubscribeCallbackMiddleware(
+                        callback_msg,
+                        callback_reply,
+                        callback_subject,
+                        callback_sid,
+                        callback_internal,
+                        serverHostname,
+                        isRequestCall
+                    )
+                );
+            };
+            clientRequest = wrappedFunction.apply(this, [subject, opts_internal, patchedCallback]);
```

Comment:
If the wrapped function throws an exception you call it twice..

Reply:
fixed.